pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -2268,10 +2271,28 @@ public void assignTableSegment(String tableNameWithType, String segmentName) {
     try {
       TableConfig tableConfig = getTableConfig(tableNameWithType);
       Preconditions.checkState(tableConfig != null, "Failed to find table config for table: " + tableNameWithType);
+
+      List<Tier> sortedTiers = null;
+      Map<String, InstancePartitions> tierToInstancePartitionsMap = null;
+
+      // Initialize tier information only in case direct tier assignment is configured
+      if (_enableTieredSegmentAssignment && CollectionUtils.isNotEmpty(tableConfig.getTierConfigsList())) {
+        sortedTiers = TierConfigUtils.getSortedTiersForStorageType(tableConfig.getTierConfigsList(),
+            TierFactory.PINOT_SERVER_STORAGE_TYPE, _helixZkManager);
+        tierToInstancePartitionsMap =
+            TierConfigUtils.getTierToInstancePartitionsMap(tableNameWithType, sortedTiers, _helixZkManager);
+
+        // Update segment tier to support direct assignment for multiple data directories
+        updateSegmentTargetTier(tableNameWithType, segmentName, sortedTiers);
```

Comment:
we can move this stmt ahead of this if-block to get a default `instancePartitionsMap`, 
```
Map<InstancePartitionsType, InstancePartitions> instancePartitionsMap =
          fetchOrComputeInstancePartitions(tableNameWithType, tableConfig);
```

then, in this if-block, we can figure out the `tierInstancePartitions` and overwrite `instancePartitionsMap` if there is a valid `tierInstancePartitions`
```
if (_enableTieredSegmentAssignment && CollectionUtils.isNotEmpty(tableConfig.getTierConfigsList())) {
...
  if (tierInstancePartitions != null) {
    instancePartitionsMap = tierInstancePartitions
  }
...
}
```